### PR TITLE
[Model] Add PaddleOCR-VL Model Support

### PR DESCRIPTION
**Pull request by yudian0504**
## Motivation

Add PaddleOCR-VL support (without paddle doc layout paser), the core implementation is based on the code from the official image (      ccr-2vdh3abv-pub.cnc.bj.baidubce.com/paddlepaddle/paddleocr-genai-vllm-server:latest  ).

Launch:
SGLANG_VLM_CACHE_SIZE_MB=2048 python -m sglang.launch_server --model-path /home/admin/PaddleOCR-VL/ --host 0.0.0.0 --port 8188 --trust-remote-code --tp-size 1  --max-running-requests 8 --mem-fraction-static 0.65  --attention-backend flashinfer --mm-attention-backend fa3 

Usage:
`
curl 'http://0:8188/v1/chat/completions' --header 'Content-Type: application/json' --data '{
        "model": "PaddleOCR-VL-0.9B",
        "messages": [
            {
                "role": "user",
                "content": [
	                {"type": "image_url", "image_url": {"url": "https://paddle-model-ecology.bj.bcebos.com/paddlex/imgs/demo_image/paddleocr_vl_demo.png"}},
    	            {"type": "text", "text": "OCR:"}
                ]
            }
        ],
        "temperature":0.0,
        "max_tokens":1000,
        "stream": false
    }'
`

More Details:
https://huggingface.co/PaddlePaddle/PaddleOCR-VL, replace vl_rec_server_url to your url.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

**Comment by albaNnaksqr** [comment]
Hi! I'm also interested in this model and have a couple of questions:
1. Do we need to register the model config in both **python/sglang/srt/hf_transformers_utils.py** and **python/sglang/srt/configs/__init__.py**? 

2. Is it neccessary to add template in **python/sglang/srt/parser/conversation.py**?
When I was working on support for another OCR model earlier, I found [clarification](https://github.com/sgl-project/sglang/pull/8778#discussion_r2258418003) like this, which seems to suggest that we don’t need to register an additional template in the code.

**Comment by yudian0504** [comment]
> Hi! I'm also interested in this model and have a couple of questions:
> 
> 1. Do we need to register the model config in both **python/sglang/srt/hf_transformers_utils.py** and **python/sglang/srt/configs/init.py**?
> 2. Is it necessary to add template in **python/sglang/srt/parser/conversation.py**?
>    When I was working on support for another OCR model earlier, I found [clarification](https://github.com/sgl-project/sglang/pull/8778#discussion_r2258418003) like this, which seems to suggest that we don’t need to register an additional template in the code.

1. I understand this is not necessary (AutoConfig + trust_remote_code is sufficient), because we don't have explicit dependencies on the config type.
2. Thank you for the reminder. I just noticed that the official team has recently updated the chat-template.jinja file (https://huggingface.co/PaddlePaddle/PaddleOCR-VL/blob/main/chat_template.jinja). My code is based on an older version of the model, but it seems that the new chat_template.jinja still has some minor issues. Therefore, adding a template can help ensure compatibility with both the old and new versions.

File: python/sglang/srt/models/paddleocr_vl.py [hunk 1]
```diff
@@ -0,0 +1,818 @@
...
+
+
+from collections.abc import Iterable
+from typing import List, Optional, Set, Tuple, Union
+
+import numpy as np
+import torch
+import torch.nn as nn
+from einops import rearrange
+from sgl_kernel.flash_attn import flash_attn_varlen_func
```
Comment (mingfeima) [inline python/sglang/srt/models/paddleocr_vl.py]: applying something like `maybe_get_vit_flash_attn_backend` in vllm : https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/paddleocr_vl.py#L616-L622 should be a better idea than directly calling `flash_attn_varlen_func` from sgl_kernel.

**Comment by Lzhang-hub** [comment]
@yudian0504 I tested the pr with your command, the result seems incorrect.
img_v3_02s6_4fa2b0e0-c4d7-4974-b4e4-9a6fbfb04a4g

File: python/sglang/srt/models/paddleocr_vl.py (hunk 1 above)
Comment (yudian0504) [reply python/sglang/srt/models/paddleocr_vl.py]: modified as you suggested

**Comment by yhyang201** [comment]
Could you rebase on main? This will fix the errors in the CI, and then it should be ready to merge.

**Comment by yudian0504** [comment]
> Could you rebase on main? This will fix the errors in the CI, and then it should be ready to merge.

done

**Review by yuan-luo — APPROVED** [review]

**Review by FlamingoPg — APPROVED** [review]

**Review by yhyang201 — APPROVED** [review]

File: python/sglang/srt/multimodal/processors/paddleocr_vlm.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+        return {
+            "mm_items": mm_items,
+            "input_ids": input_ids.tolist(),
+            "im_token_id": self.mm_tokens.image_token_id,
+            "mrope_positions": mrope_positions,
+            "mrope_position_delta": mrope_position_delta,
+        }
+
+    @staticmethod
+    def get_rope_index(
```
Comment (mickqian) [inline python/sglang/srt/multimodal/processors/paddleocr_vlm.py]: any difference with `MRotaryEmbedding.get_rope_index`?

File: python/sglang/srt/multimodal/processors/paddleocr_vlm.py [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+    )
+    image = image.resize((resized_width, resized_height))
+    return image
+
+
+async def resize_image_async(image, min_pixels, max_pixels, factor):
+    return resize_image(image, min_pixels, max_pixels, factor)
+
+
+class PaddleOCRVLImageProcessor(BaseMultimodalProcessor):
```
Comment (mickqian) [inline python/sglang/srt/multimodal/processors/paddleocr_vlm.py]: is it possible to reuse the Processor of Qwen?

File: python/sglang/srt/multimodal/processors/paddleocr_vlm.py (hunk 2 above)
Comment (yudian0504) [reply python/sglang/srt/multimodal/processors/paddleocr_vlm.py]: Okay, I'll confirm that now.

File: python/sglang/srt/multimodal/processors/paddleocr_vlm.py (hunk 3 above)
Comment (yudian0504) [reply python/sglang/srt/multimodal/processors/paddleocr_vlm.py]: I compared the implementations of both, and your suggestion is correct. The Qwen-VL processor can be used to replace most of PaddleOCR's implementation.
